compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2429,6 +2429,80 @@ TR::Register *OMR::X86::TreeEvaluator::andORStringEvaluator(TR::Node *node, TR::
    return resultReg;
    }
 
+TR::Register *OMR::X86::TreeEvaluator::OverflowCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   /*
+    *overflowCHK
+    *   =>child1(Operation)
+    *   =>child2(Operand1)
+    *   =>child3(Operand2)
+    */
+   bool firstChildEvaluatedAlready = node->getFirstChild()->getRegister()? true : false;
+   cg->evaluate(node->getFirstChild());
+   TR::Block *overflowCatchBlock = NULL;
+   TR::Instruction *prevInst = cg->comp()->getAppendInstruction();
+   // we need to do the operation again for the following cases:
+   // 1. the operation was somehow optimized to other forms, like LEA for example
+   // 2. the child was evaluated already under a different treetop
+   if (!(prevInst->getOpCode()).modifiesOverflowFlag() || firstChildEvaluatedAlready ) 
+      {
+      TR::Register *reg1; 
+      TR::Register *reg2 = cg->evaluate(node->getThirdChild());
+      TR_X86OpCodes op;
+      bool nodeIs64Bit; 
+      switch (node->getOverflowCHKInfo())
+         {
+         case TR::ladd:
+            nodeIs64Bit= true;
+            TR_ASSERT(TR::Compiler->target.is64Bit(), "overflowCHK for long add is only supported on x86 64 bits platform!!");
```

Comment:
This seems a rather aggressive limitation... it would seem to make sense to perhaps factor out part of the addsub evaluator that is run when lea folding fails and call that to do the evaluation rather than duplicating the logic. This could potentially do as a placeholder, but I think a better long term solution is preferable if possible. Thoughts?